lib/srv/forward/sshserver.go

```diff
@@ -862,28 +874,128 @@ func (s *Server) rejectChannel(chans <-chan ssh.NewChannel, errMessage string) {
 }
 
 func (s *Server) handleGlobalRequest(ctx context.Context, req *ssh.Request) {
-	// Version requests are internal Teleport requests, they should not be
-	// forwarded to the remote server.
-	if req.Type == teleport.VersionRequest {
+	switch req.Type {
+	case teleport.VersionRequest:
+		// Version requests are internal Teleport requests, they should not be
+		// forwarded to the remote server.
 		err := req.Reply(true, []byte(teleport.Version))
 		if err != nil {
 			s.log.Debugf("Failed to reply to version request: %v.", err)
 		}
 		return
+	case teleport.TCPIPForwardRequest:
+		if err := s.handleTCPIPForwardRequest(ctx, req); err != nil {
+			s.log.WithError(err).Warnf("Failed to check tcpip forward request")
+			if err := req.Reply(false, nil); err != nil {
+				s.log.Warnf("Failed to reply to global request: %v: %v", req.Type, err)
+			}
+			return
+		}
+	case teleport.CancelTCPIPForwardRequest:
+		if err := s.handleCancelTCPIPForwardRequest(ctx, req); err != nil {
+			s.log.WithError(err).Warnf("Failed to cancel tcpip forward request")
+			if err := req.Reply(false, nil); err != nil {
+				s.log.Warnf("Failed to reply to global request: %v: %v", req.Type, err)
+			}
+			return
+		}
+	default:
+		ok, payload, err := s.remoteClient.SendRequest(ctx, req.Type, req.WantReply, req.Payload)
+		if err != nil {
+			s.log.Warnf("Failed to forward global request %v: %v", req.Type, err)
+			return
+		}
+		if req.WantReply {
+			if err := req.Reply(ok, payload); err != nil {
+				s.log.Warnf("Failed to reply to global request: %v: %v", req.Type, err)
+			}
+		}
 	}
+}
 
-	ok, payload, err := s.remoteClient.SendRequest(ctx, req.Type, req.WantReply, req.Payload)
+// handleTCPIPForwardRequest handles remote port forwarding requests.
+func (s *Server) handleTCPIPForwardRequest(ctx context.Context, r *ssh.Request) error {
+	req, err := sshutils.ParseTCPIPForwardRequest(r.Payload)
 	if err != nil {
-		s.log.Warnf("Failed to forward global request %v: %v", req.Type, err)
-		return
+		return trace.Wrap(err)
 	}
 
-	if req.WantReply {
-		err = req.Reply(ok, payload)
+	// On forward server in "tcpip-forward" requests from SessionJoinPrincipal
+	//  should be rejected, otherwise it's possible to use the
+	// "-teleport-internal-join" user to bypass RBAC.
+	if s.identityContext.Login == teleport.SSHSessionJoinPrincipal {
```

Comment:
Why's the join user allowed to log in to the forward server?

Reply:
Hmm. [This commit](https://github.com/gravitational/teleport/commit/5f07b28) adds the join user check to local port forwarding, and that's what I used as a template here. But logging in as the join user is forbidden by [`Server.dispatch`](https://github.com/gravitational/teleport/blob/aa77570d65f0242acc8bc5249706d98af1bcb6bf/lib/srv/forward/sshserver.go#L1151). So maybe the check isn't necessary? 🤷

Reply:
Let's stay consistent with the rest of the code and leave the check in. 🤷